components/cdc/src/observer.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::sync::{Arc, RwLock};
+
+use raft::StateRole;
+use tikv::raftstore::coprocessor::*;
+use tikv::raftstore::Error as RaftStoreError;
+use tikv_util::collections::HashSet;
+use tikv_util::worker::Scheduler;
+
+use crate::endpoint::Task;
+use crate::Error as CdcError;
+
+/// An Observer for CDC.
+///
+/// It observes raftstore internal events, such as:
+///   1. Raft role change events,
+///   2. TODO Apply command events.
+#[derive(Clone)]
+pub struct CdcObserver {
+    sched: Scheduler<Task>,
+    // A shared registry for managing observed regions.
+    observe_regions: Arc<RwLock<HashSet<u64>>>,
+}
+
+impl CdcObserver {
+    /// Create a new `CdcObserver`.
+    ///
+    /// Events are strong ordered, so `sched` must be implemented as
+    /// a FIFO queue.
+    pub fn new(sched: Scheduler<Task>) -> CdcObserver {
+        CdcObserver {
+            sched,
+            observe_regions: Arc::default(),
+        }
+    }
+
+    /// Subscribe an region, the observer will sink events of the region into
+    /// its scheduler.
+    pub fn subscribe_region(&self, region_id: u64) {
+        self.observe_regions.write().unwrap().insert(region_id);
+    }
+
+    /// Stops observe the region.
+    pub fn unsubscribe_region(&self, region_id: u64) {
+        self.observe_regions.write().unwrap().remove(&region_id);
+    }
+
+    /// Check whether the region is subscribed or not.
+    pub fn is_subscribed(&self, region_id: u64) -> bool {
+        self.observe_regions.read().unwrap().contains(&region_id)
+    }
+}
+
+impl Coprocessor for CdcObserver {
+    fn start(&self) {}
+    fn stop(&self) {}
```

Comment:
These functions has empty implementation by default. You don't need to implement it explicitly.